docs: explain abstract classes as di tokens (#3469)
nestjs/docs.nestjs.com: content/fundamentals/dependency-injection.md

While we directly use the string `'CONNECTION'` in the above examples for illustration purposes, for clean code organization, it's best practice to define tokens in a separate file, such as `constants.ts`. Treat them much as you would symbols or enums that are defined in their own file and imported where needed.

#### Interfaces and abstract classes

> TypeScript interfaces are removed during compilation, so Nest can't reference them at runtime. This means an interface can describe the shape of a dependency, but it can't be used as a DI token by itself.

Reviewer: ```suggestion
TypeScript types/interfaces are erased during compilation, so Nest can't reference them at runtime. This means an interface can describe the shape of a dependency, but it can't be used as a DI token by itself.
```

```typescript
export interface LoggerService {
  log(message: string): void;
}

> ```

Reviewer: ```suggestion
```

just because we have this same code bellow

Since Nest resolves providers by runtime tokens, use a string or symbol token when registering a provider for an interface:

```typescript
export interface LoggerService {
  log(message: string): void;
}

export const LOGGER_SERVICE = Symbol('LOGGER_SERVICE');

@Injectable()
export class PinoLoggerService implements LoggerService {
  log(message: string) {
    // implementation details
  }
}

@Module({
  providers: [
    {
      provide: LOGGER_SERVICE,
      useClass: PinoLoggerService,
    },
  ],
})
export class AppModule {}
```

> To inject this provider, pass the token to the `@Inject()` decorator:

Reviewer: ```suggestion
To inject this provider, pass that token to the `@Inject()` decorator:
```

```typescript
@Injectable()
export class CatsService {
  constructor(
    @Inject(LOGGER_SERVICE)
    private readonly logger: LoggerService,
  ) {}
}
```

Abstract classes, unlike interfaces, exist at runtime. You can use an abstract class as both the TypeScript contract and the DI token:

```typescript
export abstract class LoggerService {
  abstract log(message: string): void;
}

@Injectable()
export class PinoLoggerService implements LoggerService {
  log(message: string) {
    // implementation details
  }
}

@Module({
  providers: [
    {
      provide: LoggerService,
      useClass: PinoLoggerService,
    },
  ],
})
export class AppModule {}
```

With an abstract class token, constructor-based injection can use the abstract class type directly and doesn't require `@Inject()`:

```typescript
@Injectable()
export class CatsService {
  constructor(private readonly logger: LoggerService) {}
}
```

> Use string or symbol tokens when the DI token should be decoupled from a class artifact, when several providers share the same interface, or when exposing an abstract class as part of your API isn't desirable. Use abstract classes when one artifact should act as both the contract and the runtime token, and you prefer simpler constructor injection. A plain interface is still a good choice when the type is only used for compile-time checking and no DI token is needed.

Reviewer: I guess it would be good to clarify why we are telling to use JS symbols as a token

Author: Good point. I’ll clarify that symbols provide a unique runtime identity, which helps prevent accidental token collisions compared with string tokens, especially in libraries or larger applications. I’ll also mention that the same exported symbol instance must be reused for registration and injection.
